codelab-master/apps/playground/src/app/quiz/shared/services/quiz.service.ts

```diff
@@ -0,0 +1,217 @@
+  quizData: Quiz[] = JSON.parse(JSON.stringify(QUIZ_DATA));
+  question: QuizQuestion;
+  questions: QuizQuestion[];
+  answers: number[];
+  multipleAnswer: boolean;
+  totalQuestions: number;
+  currentQuestionIndex = 1;
+
+  quizId: string;
+  startedQuizId: string;
+  continueQuizId: string;
+  completedQuizId: string;
+  indexOfQuizId: number;
+  quizCompleted: boolean;
+  status: string;
+
+  correctAnswers = [];
+  correctAnswersForEachQuestion = [];
+  correctAnswerOptions: number[] = [];
+  numberOfCorrectAnswers: number;
+  correctAnswersCountSubject = new BehaviorSubject<number>(0);
+
+  userAnswers = [];
+  previousUserAnswers: any[] = [];
+  previousUserAnswersText: any[] = [];
+  previousUserAnswersInnerText = [];
+  previousUserAnswersTextSingleAnswer: string[] = [];
+  previousUserAnswersTextMultipleAnswer: string[] = [];
+
+  explanation: string;
+  explanationText: string;
+  correctOptions: string;
+  correctMessage: string;
+
+  isAnswered: boolean;
+  checkedShuffle: boolean;
+
+  correctSound = new Howl({
+    src: '../../../assets/audio/sound-correct.mp3',
+    html5: true,
+    format: ['mp3', 'aac']
+  });
+  incorrectSound = new Howl({
+    src: '../../../assets/audio/sound-incorrect.mp3',
+    html5: true,
+    format: ['mp3', 'aac']
+  });
+
+
+  constructor(
+    private activatedRoute: ActivatedRoute,
+    private router: Router
+  ) {
+    this.quizId = this.activatedRoute.snapshot.paramMap.get('quizId');
+    this.indexOfQuizId = this.quizData.findIndex(el => el.quizId === this.quizId);
+  }
+
+  getCorrectAnswers(question: QuizQuestion) {
+    if (this.question) {
+      const identifiedCorrectAnswers = question.options.filter((option) => option.correct);
+      this.numberOfCorrectAnswers = identifiedCorrectAnswers.length;
+      this.correctAnswerOptions = identifiedCorrectAnswers.map((option) => question.options.indexOf(option) + 1);
+
+      this.correctAnswersForEachQuestion.push(this.correctAnswerOptions);
+      this.correctAnswers.push(this.correctAnswersForEachQuestion);
+
+      this.setCorrectAnswerMessagesAndExplanationText(this.correctAnswersForEachQuestion.sort());
+      return identifiedCorrectAnswers;
+    }
+  }
+
+  // shuffle questions array in-place using Durstenfeld's shuffling algorithm
+  shuffleQuestions(questions: QuizQuestion[]): void {
+    for (let i = questions.length - 1; i >= 0; i--) {
+      const j = Math.floor(Math.random() * (i + 1));
+      [questions[i], questions[j]] = [questions[j], questions[i]];
+    }
+  }
+
+  // shuffle answers array in-place using Durstenfeld's shuffling algorithm
+  shuffleAnswers(answers: Option[]): void {
+    for (let i = answers.length - 1; i >= 0; i--) {
+      const j = Math.floor(Math.random() * (i + 1));
+      [answers[i], answers[j]] = [answers[j], answers[i]];
+    }
+  }
+
+  /********* setter functions ***********/
+  setCorrectAnswerMessagesAndExplanationText(correctAnswers: number[]): void {
+    if (correctAnswers[0][0]) {
```

Comment:
please use a loop here.

Reply:
I added a for loop. Should it look like this? 

setCorrectAnswerMessagesAndExplanationText(correctAnswers: number[]): void {
    for (let i = 0, j = 0; i = 0, j < correctAnswers.length; j++) {
      if (correctAnswers[i][j]) {
        this.correctOptions = correctAnswers[0][j];
        this.correctMessage = 'The correct answer was Option ' + this.correctOptions + '.';
      }
      if (correctAnswers[i][j] && correctAnswers[0][j + 1]) {
        this.correctOptions = correctAnswers[0][0].toString().concat(' and ', correctAnswers[0][j + 1]);
        this.correctMessage = 'The correct answers were Options ' + this.correctOptions + '.';
      }
      if (correctAnswers[i][j] && correctAnswers[0][j + 1] && correctAnswers[0][j + 2]) {
        this.correctOptions = correctAnswers[0][0].toString().concat(', ', correctAnswers[0][1], ' and ', correctAnswers[0][2]);
        this.correctMessage = 'The correct answers were Options ' + this.correctOptions + '.';
      }
      if (correctAnswers[i][j] && correctAnswers[0][j + 1] && correctAnswers[0][j + 2] && correctAnswers[0][j + 3]) {
        this.explanationText = 'All were correct!';
        this.correctMessage = 'All were correct!';
      }
    }

Reply:
Okay, I just tested it and it seems to be working fine.

Reply:
OK I tested this function and it seems to be working fine.

Reply:
I've got the correctOptions text working (refactored the function to use a for loop as suggested), but the explanationText doesn't seem to display, getting undefined in the console in the QuizComponent. I tried with a getter and using dependency injection, doesn't seem to want to get the value of explanationText from the QuizService.

Reply:
trying this.explanationText = this.quizData[this.indexOfQuizId].questions[this.quizService.currentQuestionIndex].explanation; in the QuizComponent, explanation text displays in the console, but doesn't seem to data-bind to the template...

Reply:
If I remove the *ngIf="isAnswered()" from the span, the explanationText will display.